File: modules/widgets/lib/rin/widgets/content/WImage.cpp
#include "rin/widgets/content/WImage.hpp"
#include "rin/graphics/DeviceImage.hpp"
#include "rin/graphics/ResourceManager.hpp"
#include "rin/widgets/graphics/SimpleBatchedDrawCommand.hpp"
#include "rin/widgets/graphics/WidgetDrawCommands.hpp"

void WImage::SetTextureId(const int& textureId)
{
    auto isNew = _textureId != textureId;
    _textureId = textureId;
    if (isNew)
    {
        CheckSize();
    }
}

int WImage::GetTextureId() const
{
    return _textureId;
}

void WImage::SetBorderRadius(const Vec4<float>& borderRadius)
{
    _borderRadius = borderRadius;
}

Vec4<float> WImage::GetBorderRadius() const
{
    return _borderRadius;
}

void WImage::SetTint(const Vec4<float>& tint)
{
    _tint = tint;
}

Vec4<float> WImage::GetTint() const
{
    return _tint;
}

void WImage::CollectContent(const TransformInfo& transform, WidgetDrawCommands& drawCommands)
{
    if (_textureId < 0)
    {
        drawCommands.Add(
            SimpleBatchedDrawCommand::Builder{}
            .AddRect(
                GetContentSize(),
                transform.transform,
                GetBorderRadius(),
                GetTint()
            )
            .Finish()
        );
    }
    else
    {
        drawCommands.Add(
            SimpleBatchedDrawCommand::Builder{}
            .AddTexture(
                _textureId,
                GetContentSize(),
                transform.transform,
                GetBorderRadius(),
                GetTint()
            )
            .Finish()
        );
    }
}

Vec2<float> WImage::ComputeContentSize()
{
    if (auto vkImage = GraphicsModule::Get()->GetResourceManager()->GetTextureImage(GetTextureId()))
    {
        auto extent = vkImage->GetExtent();
        return {static_cast<float>(extent.width), static_cast<float>(extent.height)};
    }
    return {0.0f, 0.0f};
}

File: aerox.Runtime.Widgets/Content/WImage.cs
using System.Runtime.InteropServices;
using aerox.Runtime.Graphics;
using aerox.Runtime.Math;
using aerox.Runtime.Widgets.Graphics;
using aerox.Runtime.Widgets.Graphics.Quads;

namespace aerox.Runtime.Widgets.Content;

/// <summary>
///     Draw's a <see cref="Texture" /> if provided or a colored rectangle. Supports tint.
/// </summary>
public class WImage : Widget
{

    private int _textureId = -1;

    public WImage()
    {
    }

    public int TextureId
    {
        get => _textureId;
        set
        {
            _textureId = value;
            TryUpdateDesiredSize();
        }
    }

    public Color Tint { get; set; } = new Color(1.0f);

    public Vector4<float> BorderRadius { get; set; } = new(0.0f);

    protected override void OnDispose(bool isManual)
    {
        base.OnDispose(isManual);
    }
    
    protected override Size2d ComputeDesiredContentSize()
    {
        if (SGraphicsModule.Get().GetResourceManager()
                .GetTextureImage(TextureId) is { } texture)
        {
            return new Size2d
            {
                Width = texture.Extent.width,
                Height = texture.Extent.height
            };
        }

        return new Size2d();
    }
    
    protected override void OnRemovedFromSurface(Surface surface)
    {
        base.OnRemovedFromSurface(surface);
    }

    public override void CollectContent(TransformInfo info, DrawCommands drawCommands)
    {
        //throw new NotImplementedException();
        if (TextureId != -1)
        {
            drawCommands.Add(new QuadDrawCommand([new Quad(GetContentSize(), info.Transform)
            {
                TextureId = TextureId,
                Color = Tint,
                BorderRadius = BorderRadius
            }]));
        }
    }
}